### PR TITLE
Support for undirected segments.

### PR DESCRIPTION
**Pull request by mtyaka**
This patch adds some support for defining and grading segments.
Segments work just like vectors except that they are drawn without the arrow.

Segments are currently defined as special types of vectors. To define a segment,
use `type: "segment"` in the `vectors: [...]` definition.

Segments work with all the same checks as vectors. Two new types of checks
intended to be used specifically with segments have been added:

* `segment_angle` - it works just like `angle`, except that the segment can
  be drawn in either direction (`segment_angle: 0` is equivalent to
  `segment_angle: 180`, etc).
* `segment_coords`: This defines the expected coordinates of the segment
  ending points, for example: `[[1.5, 1.5], [3, 3]]`. Since segments aren't
  directed, `segment_coords: [[x1, y1], [x2, y2]]` is equivalent to
  `[[x2, y2], [x1, y1]]`.

### CODE REVIEW

**Comment by antoviaque** [comment]
@dseaton FYI - this hasn't gone through code review and testing yet, but in case you want to have an early look.

File: vectordraw.py [hunk 1]
```diff
@@ -48,7 +70,41 @@ def check_tip_x(check, vectors):
...
+    for expected_coords, vec_coords in ([expected[0], vec.tail], [expected[1], vec.tip]):
+        if expected_coords[0] == '_':
+            if expected_coords[1] != '_':
+                # Only y given; not interested in x.
+                if abs(expected_coords[1] - vec_coords.y) > tolerance:
+                    return True
+        elif expected_coords[1] == '_':
+            # Only x given; not interested in y.
+            if abs(expected_coords[0] - vec_coords.x) > tolerance:
+                return True
```
Comment (e-kolpakov) [inline vectordraw.py]: @mtyaka most likely it could be achieved in a more elegant way

    def check_coord(expected, actual):
        return expected == '_' or abs(expected - actual) > tolerance
    
    if not (check_coord(expected_coords[0], vec_coords.x) and check_coord(expected_coords[1], vec_coords.x):
        return True  # do I understand right that this function returns True if check fails?

    if expected_coords[0] != '_' and expected_coords[1] != '_':
        # same as below

The trick is, distance check is a stronger assumption, so if it holds above checks will always hold as well. Basically, checking for both coordinates be in `tolerance` distance of target point independently forms a square; checking for distance from target point forms an inscribed circle.

File: vectordraw.py [hunk 2]
```diff
@@ -48,7 +70,41 @@ def check_tip_x(check, vectors):
...
+        else:
+            # Both coordinates given; check distance from expected point.
+            x = expected_coords[0] - vec_coords.x
+            y = expected_coords[1] - vec_coords.y
+            if math.hypot(x, y) > tolerance:
+                return True
+    return False
```
Comment (e-kolpakov) [inline vectordraw.py]: @mtyaka nit: I find counter-intuitive that this function returns False if the check pass. Maybe it should do the reverse of it. Or have a better name. Or both.

File: vectordraw.py [hunk 3]
```diff
@@ -67,9 +120,25 @@ def check_angle(check, vectors):
     y = vec.tip.y - vec.tail.y
     dot_product = x * math.cos(expected) + y * math.sin(expected)
     angle = math.degrees(math.acos(dot_product / vec.length))
-    if abs(angle) > tolerance:
+    return abs(angle) > tolerance
```
Comment (e-kolpakov) [inline vectordraw.py]: @mtyaka nit: same here - I find it counter-intuitive that this function returns False if the check pass.

File: vectordraw.py [hunk 4]
```diff
@@ -67,9 +120,25 @@ def check_angle(check, vectors):
...
         return 'The angle of {} is incorrect. Your angle: {:.1f}'.format(vec.name, vec.angle)
 
+def check_segment_angle(check, vectors):
+    # Segments are not directed, so we must check the angle between the segment and
+    # the vector that represents it, as well as its opposite vector.
+    vec = vectors[check['vector']]
+    tolerance = check.get('tolerance', 2.0)
+    expected = math.radians(check['expected'])
+    if _check_angle(vec, expected, tolerance):
+        if _check_angle(vec.opposite(), expected, tolerance):
```
Comment (e-kolpakov) [inline vectordraw.py]: @mtyaka nit: I would still use `and`. If you're concerned of `\`, maybe formatting it like this would look better?

    if (
        _check_angle(vec, expected, tolerance) and
        _check_angle(vec.opposite(), expected, tolerance)
    ):

File: api-example.xml [hunk 5]
```diff
@@ -1,149 +1,219 @@
...
+            # Only x given; not interested in y.
+            if abs(expected_coords[0] - vec_coords.x) > tolerance:
+                return True
+        else:
+            # Both coordinates given; check distance from expected point.
+            x = expected_coords[0] - vec_coords.x
+            y = expected_coords[1] - vec_coords.y
+            if math.hypot(x, y) > tolerance:
+                return True
+    return False
```
Comment (e-kolpakov) [inline api-example.xml]: @mtyaka see comments in vectordraw.py

File: api-example.xml [hunk 6]
```diff
@@ -1,149 +1,219 @@
...
+
+def check_coords(check, vectors):
+    vec = vectors[check['vector']]
+    expected = check['expected']
+    tolerance = check.get('tolerance', 1.0)
```
Comment (e-kolpakov) [inline api-example.xml]: @mtyaka it might make sense to give this tolerance a better name: `coord_tolerance` or something. We'll have to keep it backward compatible though.

**Comment by e-kolpakov** [comment]
@mtyaka there're a couple of improvement suggestions, but overall looks good and works just fine. :+1: conditional notes are addressed.

**Comment by dseaton** [comment]
@mtyaka @antoviaque Happy to try this and share a quick example. Thoughts?

File: vectordraw.py [hunk 7]
```diff
@@ -2,6 +2,28 @@
 # script tag of the problem XML definition - see api-example.xml for
 # an example.
 
+import json
+import math
+
+
+try:
+    custom_checks
```
Comment (smarnach) [inline vectordraw.py]: How could this possibly be defined at this point?  Can't we just have `custom_checks = {}` here with a comment that this can be customised?  Same thing applies to `success_message`.

File: vectordraw.py [hunk 8]
```diff
@@ -2,6 +2,28 @@
...
+try:
+    success_message
+except NameError:
+    success_message = 'Good job!'
+
+
+def vglcfn(e, ans):
+    """Main grading function."""
+    global custom_checks
+    global success_message
```
Comment (smarnach) [inline vectordraw.py]: No need to declare these as globals – you don't assign to these names.

File: vectordraw.py (hunk 1 above)
Comment (smarnach) [inline vectordraw.py]: I'm confused by `False` meaning "correct" and `True` meaning "not correct" as well.

Yet another option how to implement this more elegantly:

    def _coord_delta(expected, actual):
        if expected == '_':
            return 0.0
        return expected - actual

    def _check_coords(vec, expected, tolerance):
        for expected_coords, vec_coords in ([expected[0], vec.tail], [expected[1], vec.tip]):
            delta_x = _coord_delta(expected_coords[0], vec_coords.x)
            delta_y = _coord_delta(expected_coords[1], vec_coords.y)
            if math.hypot(delta_x, delta_y) > tolerance:
                return True
        return False

Basically, set the error to zero for dimensions we don't care about.

File: vectordraw.py (hunk 4 above)
Comment (smarnach) [inline vectordraw.py]: @e-kolpakov The downside of this code layout is that the last line looks a bit sad.  Otherwise agreed. :)

File: vectordraw.py [hunk 9]
```diff
@@ -48,7 +70,41 @@ def check_tip_x(check, vectors):
...
+    if _check_coords(vec, expected, tolerance):
+        return 'Vector {} coordinates are not correct.'.format(vec.name)
+
+def check_segment_coords(check, vectors):
+    vec = vectors[check['vector']]
+    expected = check['expected']
+    tolerance = check.get('tolerance', 1.0)
+    if _check_coords(vec, expected, tolerance) and \
+            _check_coords(vec.opposite(), expected, tolerance):
+        return 'Segment {} coordinates are not correct.'.format(vec.name)
```
Comment (smarnach) [inline vectordraw.py]: An alternative to checking the vector and its opposite is to normalise the direction of segments – both the expected values as well as the vectors entered by the student.  An easy way to do so would be to lexicographically sort the points.
Your approach is completely fine, no need to change anything.  You might have gotten away with slightly less work by subclassing `Vector` with a `Segment` class that normalises the point order, but you definitely won't save any work by doing it now. :)

**Comment by smarnach** [comment]
@mtyaka Looks good to me, :+1:, pending the minor comments.

**Comment by antoviaque** [comment]
> Happy to try this and share a quick example. Thoughts?

@dseaton Definitely - it can be useful to confirm that it works correctly for the use-case you have in mind, testing it on a quick example can help making sure we have the needs covered.

File: vectordraw.py (hunk 1 above)
Comment (mtyaka) [inline vectordraw.py]: @e-kolpakov @smarnach Yeah, the `False` as correct was confusing. I reversed the meanings of `False` and `True` to what you'd normally expect and renamed the function so that there's no confusion with the other `check_*` functions which return a truthy value (the error message) on error.

I also replaced the messy nested-if code with the neat `coord_delta` approach, thanks for the suggestions.

File: vectordraw.py (hunk 2 above)
Comment (mtyaka) [inline vectordraw.py]: Yup, both :)

File: vectordraw.py (hunk 3 above)
Comment (mtyaka) [inline vectordraw.py]: Reversed the return value & renamed.

File: vectordraw.py (hunk 4 above)
Comment (mtyaka) [inline vectordraw.py]: @e-kolpakov Thanks, that looks nicer indeed.

File: api-example.xml (hunk 6 above)
Comment (mtyaka) [inline api-example.xml]: @e-kolpakov When you define the tolerance in the `expected_result` setting, you already have to pass it as `coord_tolerance`. Same goes for all other checks (`angle_tolerance`, `tail_tolerance`, `length_tolerance`...). They're all converted to just `tolerance` inside `vectordraw.js` code when converting the `expected_result` entry to a list of checks.

File: vectordraw.py (hunk 7 above)
Comment (mtyaka) [inline vectordraw.py]: @smarnach You're right, that was unnecessary. I removed the ugly `try-except` blocks and just added a comment that these can be modified by the end user directly.

File: vectordraw.py (hunk 9 above)
Comment (mtyaka) [inline vectordraw.py]: @smarnach I actually started out by creating a `class Segment(Vector)`. But since conceptually a vector is a specialized version of a segment (segment coords + direction) - not the other way around - that code just bothered me too much and I settled for `def opposite` instead...

**Comment by dseaton** [comment]
@smarnach @antoviaque I have created a demo in the assessment course.
https://studio.edge.edx.org/container/i4x://DavidsonNext/Ann101/vertical/2e83da6c4be04aeb957120497757f30c?action=new

Note, I have checked out the segments branch and created an example. Should I simply push to segments?

**Comment by antoviaque** [comment]
@dseaton Thanks!

For pushing the code, it's usually best to have only one person handling a branch at a time, and handle the merges through PRs -- this way there is no conflict at the moment of pushing. In cases where the branch you depend on isn't merged, you can create another branch based on it, and target the PR against the branch you depend on instead of master.

In the current case, since it's now merged on master, you can create a PR the usual way - if the example doesn't need a review, you can directly merge the PR; it's still useful to use the PR to be able to notice the merge, but feel free to add examples to the repo directly, they are pretty self-contained.

**Comment by dseaton** [comment]
@antoviaque @smarnach Got it. I have already committed one example to the "segments" branch. Will commit and make a pull request, plus add feedback there.
